src/error.rs

```diff
@@ -17,109 +19,127 @@ struct ErrorImpl {
     cause: Option<Cause>,
 }
 
-#[derive(Debug, PartialEq)]
-pub(super) enum Kind {
+/// Represents the kind of the error.
+///
+/// This enum is non-exhaustive.
+#[non_exhaustive]
+pub enum Kind {
+    /// Error occured while parsing.
     Parse(Parse),
+    /// Error occured while executing user code.
     User(User),
     /// A message reached EOF, but is not complete.
-    IncompleteMessage,
+    IncompleteMessage(Sealed),
     /// A connection received a message (or bytes) when not waiting for one.
     #[cfg(feature = "http1")]
-    UnexpectedMessage,
+    UnexpectedMessage(Sealed),
     /// A pending item was dropped before ever being processed.
-    Canceled,
+    Canceled(Sealed),
     /// Indicates a channel (client or body sender) is closed.
-    ChannelClosed,
+    ChannelClosed(Sealed),
     /// An `io::Error` that occurred while trying to read or write to a network stream.
     #[cfg(any(feature = "http1", feature = "http2"))]
-    Io,
+    Io(Sealed),
     /// Error occurred while connecting.
-    Connect,
+    Connect(Sealed),
     /// Error creating a TcpListener.
     #[cfg(all(
         any(feature = "http1", feature = "http2"),
         feature = "tcp",
         feature = "server"
     ))]
-    Listen,
+    Listen(Sealed),
     /// Error accepting on an Incoming stream.
     #[cfg(any(feature = "http1", feature = "http2"))]
     #[cfg(feature = "server")]
-    Accept,
+    Accept(Sealed),
     /// Error while reading a body from connection.
     #[cfg(any(feature = "http1", feature = "http2", feature = "stream"))]
-    Body,
+    Body(Sealed),
```

Comment:
I kinda think these various body errors probably could be organized into a single enum. The variant names by themselves right now aren't super clear, I think...